### PR TITLE
NdlcLayoutRenderer - Nested Diagnostics Logical Context

### LINKED ISSUE #1615: Add context on logger level
Feature:

We like to do:

``` c#
var logger = LogManager.GetCurrentClassLogger();
logger.AddContext("declarationId", 10);
```

rendering, something like this:

LDC? "logger diagnostic context"

```
${ldc}
```

### PR DESCRIPTION
**Pull request by snakefoot**
Make it easier to support Microsoft ILogger BeginScope (And maybe resolve #1615)

Maybe rename to Nested Diagnostics Logical Context (NDLC) ?

### CODE REVIEW

**Comment by 304NotModified** [comment]
Cool! Great!

I propose:

- the class name for the context: `NestedDiagnosticsLogicalContext ` (consistent with `NestedDiagnosticsContext `, `MappedDiagnosticsContext` and `MappedDiagnosticsLogicalContext`)

- the class name and renderer name for the layout renderer: NdlcLayoutRender ${NDLC} (consistent with NdcLayoutRenderer/${NDC}, MDC, MDLC etc)

**Comment by snakefoot** [comment]
> I propose:

@304NotModified Accepted

**Comment by 304NotModified** [comment]
thanks!

File: src/NLog/NestedDiagnosticsLogicalContext.cs [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+            return current;
+        }
+
+        /// <summary>
+        /// Pops the top message off the current stack
+        /// </summary>
+        /// <returns>The top message which is no longer on the stack.</returns>
+        public static object Pop()
+        {
+            var current = GetThreadLocal();
```
Comment (304NotModified) [inline src/NLog/NestedDiagnosticsLogicalContext.cs]: @snakefoot 

see it now in the code. It this really a Pop (should remove it from the stack)? looks like a Top method

Comment (snakefoot) [inline src/NLog/NestedDiagnosticsLogicalContext.cs]: Yes the XML-comment is wrong. Pop removes from the stack.

Comment (snakefoot) [inline src/NLog/NestedDiagnosticsLogicalContext.cs]: Sorry not sure I understand the Pop-method modifies the stack by removing the top-elemente and returning it.

**Comment by 304NotModified** [comment]
@snakefoot could you please help me by adding some docs to the wiki? Thanks!

**Comment by snakefoot** [comment]
@304NotModified Added new wiki page: https://github.com/NLog/NLog/wiki/NDLC-Layout-Renderer

**Comment by 304NotModified** [comment]
thanks!!

**Comment by furier** [comment]
Could the documentation possibly be included to explain what is the purpose of NDLC and NDC and maybe how it differs from MDLC and MDC? With NLog 4.5 structured logging MDLC seems to add metadata properties that are part of the structured data being logged but not part of the message rendered, correct? What does NDLC do then? append the pushed value to the rendered message or prepended, does it also support structured data, etc...?

**Comment by 304NotModified** [comment]
@furier please create a new github issue. Closed issues are difficult to track. Thanks!

**Pull request merged**